### PR TITLE
Add library to facilitate SQL backup export

### PR DESCRIPTION
**Pull request by aswasif007**
## Description

In this PR, we are adding library to facilitate exporting the latest SQL backup of an application environment. It can be done by attaching the following piece of code to a command:

```javascript
const exportCommand = new SQLExportCommand( app, env, outputPath );
await exportCommand.runSequence();
```
An object of the `SQLExportCommand` class acts as a show-runner for exporting an SQL backup. It runs the following in sequence:
- Initiates export generation for the latest backup
- Generates a pre-signed URL for the exported file
- Downloads the export to the specified path

The implementation also accounts for the following cases:
- There is already an export running for the latest backup
- An export for the latest backup already exists

The implementation has assumed the following limitations, which will be addressed in subsequent work:
- Doesn't check for permission (and rely on the API to fail if the user does not have sufficient permission)
- Doesn't allow user to select network-sites
- Doesn't allow user to make table selection

This class is to be used in the upcoming `vip dev-env sync sql` and `vip export sql` commands. 

## Steps to Test

To manually test this change, we need to add a command first.

1. Add a new file `src/bin/vip-export-sql.js` and paste the following code:
```js
import command from 'lib/cli/command';
import { SQLExportCommand } from 'lib/sql-export';

const examples = [];

const appQuery = `
  id,
  name,
  type,
  organization { id, name },
  environments{
    id
    appId
    type
    name
    primaryDomain { name }
  }
`;

command( {
  appContext: true,
  appQuery,
  envContext: true,
  requiredArgs: 0,
  module: 'export-sql',
} )
  .option(
    'output',
    'Specify the location where you want to save the export file',
  )
  .examples( examples )
  .argv( process.argv, async ( arg: string[], { app, env } ) => {
    const exportCommand = new SQLExportCommand( app, env );
    await exportCommand.runSequence();
  } );

```
2. Add `"vip-export-sql": "dist/bin/vip-export-sql.js",` line to `package.json#bin`
3. Add `.command( 'export-sql', 'Export SQL files ' )` line [here](https://github.com/Automattic/vip-cli/blob/8a068a7560c845ad377bd913d048287cd9533885/src/bin/vip.js#L46)
4. Run `rm -rf dist && npm link node ./dist/bin/vip export-sql @<app>.<env>` (if you are an automattician and using proxy, you may have to add `VIP_PROXY=socks://localhost:8080`)

### CODE REVIEW

File: src/lib/common.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Polls a function until its return value satisfies a condition
+ * @param       {function} fn A function to poll
+ * @param       {int} interval Poll interval in milliseconds
+ * @param       {function} isDone A function that returns true if the polling should stop
+ * @resolves		{Promise} A promise which resolves when the polling is done
+ */
+export async function pollUntil( fn, interval, isDone ) {
+	return new Promise( async ( resolve, reject ) => {
```
Comment (abdullah-kasim) [inline src/lib/common.js]: You'll want `return await new Promise` here. See [ref under "Code example Anti-Pattern: Returning a promise without tagging the function as async"](https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/errorhandling/returningpromises.md#code-example-anti-pattern-returning-a-promise-without-tagging-the-function-as-async), pretty surprising but that's how it is.

File: src/lib/sql-export/export-command.js [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	 * @param {object} env The environment object
+	 * @param {string} outputFile The output file path
+	 * @constructor
+	 */
+	constructor( app, env, outputFile ) {
+		this.app = app;
+		this.env = env;
+		this.outputFile = outputFile;
+		this.progressTracker = new ProgressTracker( [
+			{ id: 'prepare', name: 'Preparing' },
```
Comment (abdullah-kasim) [inline src/lib/sql-export/export-command.js]: nit: Maybe we should use constants for the id? i.e. `const PROGRESS_PREPARE = 'prepare'` defined somewhere in this file. Helps future readers that aren't familiar with `ProgressTracker` and minimize typo chances.

File: src/lib/common.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Polls a function until its return value satisfies a condition
+ * @param       {function} fn A function to poll
+ * @param       {int} interval Poll interval in milliseconds
+ * @param       {function} isDone A function that returns true if the polling should stop
```
Comment (abdullah-kasim) [inline src/lib/common.js]: nit:

```suggestion
 * @param       {function} isDone A function that accepts the return of `fn`. Stops the polling if it returns true
```

**Review by abdullah-kasim — APPROVED** [review]
I don't really see any issue, though I'm not too familiar with `vip-cli` myself.

File: src/lib/common.js (hunk 1 above)
Comment (aswasif007) [reply src/lib/common.js]: Wow! TIL

@Automattic/vip-platform-forno

File: src/lib/sql-export/export-command.js [hunk 4]
```diff
@@ -0,0 +1,337 @@
+/**
```
Comment (saroshaga) [inline src/lib/sql-export/export-command.js]: No need to name it as `export-command` , I think `sql-export.js` should suffice to keep in line with other modules

File: src/lib/common.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+/**
```
Comment (saroshaga) [inline src/lib/common.js]: probably `src/lib/utils.js`

File: src/lib/sql-export/export-command.js [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+	if ( ! success ) {
+		throw new Error();
+	}
+}
+
+/**
+ * Class representing an export command workflow
+ */
+export class ExportCommand {
```
Comment (saroshaga) [inline src/lib/sql-export/export-command.js]:
```suggestion
export class SQlExportCommand {
```
or 

```suggestion
export class DatabaseExportCommand {
```

File: src/lib/common.js (hunk 1 above)
Comment (saroshaga) [reply src/lib/common.js]: Would it be cleaner to just not explicitly return a new Promise?

File: src/lib/sql-export/export-command.js (hunk 6 above)
Comment (saroshaga) [reply src/lib/sql-export/export-command.js]: If we're using `DatabaseExport` then we should reflect that everywhere else

**Review by saroshaga — COMMENTED** [review]
Looks good!! You ended up building most of the command 😄 

Left a few comments but they are minor

File: src/lib/common.js (hunk 1 above)
Comment (sjinks) [reply src/lib/common.js]: See also: https://eslint.org/docs/latest/rules/no-async-promise-executor

I think if you update the branch, ESLint will complain.

Comment (aswasif007) [reply src/lib/common.js]: > Would it be cleaner to just not explicitly return a new Promise?

That's a very good idea. Wonder how I missed that :smile:

**Comment by sjinks** [comment]
@aswasif007 could you please address these issues, as they break CI: https://github.com/Automattic/vip-cli/actions/runs/3949306457?

Thank you.

**Comment by aswasif007** [comment]
Thanks for flagging @sjinks : https://github.com/Automattic/vip-cli/pull/1251

**Pull request merged**